### PR TITLE
Patchified corrdiff integration

### PR DESCRIPTION
**Pull request by tge25**
# Modulus Pull Request

## Description
Patchified training and inference code integration for corrdiff
Flexible patch generation with customized overfitting and boundary crop
Currently patchified inference only works for stochastic sampler
Provide example commands for patchified training and generation

## Checklist

- [X] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

**Review by MortezaMardani — APPROVED** [review]

File: examples/generative/corrdiff/conf/config_generate.yaml [hunk 1]
```diff
@@ -18,7 +18,7 @@ hydra:
 
 
 ## Main options
```
Comment (mnabian) [inline examples/generative/corrdiff/conf/config_generate.yaml:20]: Some of the default configs are changed here, including the batch size and validation times. Is that necessary?

File: examples/generative/corrdiff/conf/config_train.yaml [hunk 2]
```diff
@@ -18,6 +18,7 @@ hydra:
```
Comment (mnabian) [inline examples/generative/corrdiff/conf/config_train.yaml:18]: Same here. Please keep the default configs as is where possible.

File: examples/generative/corrdiff/generate.py [hunk 3]
```diff
@@ -469,15 +481,6 @@ def generate(
     "Elucidating the Design Space of Diffusion-Based Generative Models".
     """
 
-    if sampling_method == "stochastic":
```
Comment (mnabian) [inline examples/generative/corrdiff/generate.py:472]: Perhaps we need to revert this change?

File: examples/generative/corrdiff/train.py [hunk 4]
```diff
@@ -26,6 +26,7 @@
 
 import hydra
 import torch
+import modulus
```
Comment (mnabian) [inline examples/generative/corrdiff/train.py:29]: Do we need this import?

File: examples/generative/corrdiff/training/training_loop.py [hunk 5]
```diff
@@ -181,9 +186,22 @@ def training_loop(
         logger0.success("Loaded the pre-trained regression network")
     else:
         interface_kwargs = {}
-    loss_fn = construct_class_by_name(**loss_kwargs, **interface_kwargs)
+
+    if (loss_kwargs.class_name == "modulus.metrics.diffusion.ResLoss"):
+        loss_fn = construct_class_by_name(**loss_kwargs, **interface_kwargs,
+        img_shape_x = img_shape_x,
```
Comment (mnabian) [inline examples/generative/corrdiff/training/training_loop.py:192]: I suggest we include these kwargs as part of the interface_kwargs. Then `construct_class_by_name(**loss_kwargs, **interface_kwargs` will work for any loss.

File: examples/generative/corrdiff/training/training_loop.py [hunk 6]
```diff
@@ -262,6 +283,8 @@ def training_loop(
                 )
                 training_stats.report("Loss/loss", loss)
                 loss.sum().mul(loss_scaling / batch_gpu_total).backward()
+                all_loss.append(loss.sum().mul(loss_scaling / batch_gpu_total).item())
```
Comment (mnabian) [inline examples/generative/corrdiff/training/training_loop.py:286]: Loss reduction is done twice. Can we instead have
`loss = loss.sum().mul(loss_scaling / batch_gpu_total)
loss.backward()
all_loss.append(loss.time())
`

File: examples/generative/corrdiff/train.py [hunk 7]
```diff
@@ -90,6 +90,9 @@ def main(cfg: DictConfig) -> None:
     c.out_channels = getattr(cfg, "out_channels", [0, 17, 18, 19])
     c.img_shape_x = getattr(cfg, "img_shape_x", 448)
     c.img_shape_y = getattr(cfg, "img_shape_y", 448)
+    c.patch_shape_x = getattr(cfg, "patch_shape_x", 448)
+    c.patch_shape_y = getattr(cfg, "patch_shape_y", 448)
+    c.patch_num = getattr(cfg, "patch_num", 1)
```
Comment (mnabian) [inline examples/generative/corrdiff/train.py:95]: Could `patch_num` be inferred from `path_shape` instead of making it configurable?

File: examples/generative/corrdiff/training/training_loop.py [hunk 8]
```diff
@@ -154,6 +157,8 @@ def training_loop(
     )
 
     img_in_channels = len(in_channels)  # noise + low-res input
+    if img_shape_x!=patch_shape_x or img_shape_y!=patch_shape_y:
+        img_in_channels *= 2
```
Comment (mnabian) [inline examples/generative/corrdiff/training/training_loop.py:161]: This is not clear to me.Would be great if you can please add a comment next to this and explain it

**Comment by mnabian** [comment]
@tge25 thanks for the PR! Could you please run the inference code without patching on the light validation set and report the results? This is to make sure changes do not negatively affect the baseline code.

File: examples/generative/corrdiff/train.py (hunk 7 above)
Comment (tge25) [reply examples/generative/corrdiff/train.py:95]: It's quite challenging. The selection of patch_num is highly dependent on GPU memory, patch shape, batch size, and other factors to optimize memory usage effectively. Determining the optimal value might require extensive testing across various scenarios. Therefore, I think setting it as a tunable parameter might be a good idea?

**Comment by mnabian** [comment]
@tge25 could you please resolve the conflicts?

File: examples/generative/corrdiff/generate.py [hunk 9]
```diff
@@ -117,7 +125,10 @@ def main(cfg: DictConfig) -> None:
         }
     else:
         raise ValueError(f"Unknown sampling method {sampling_method}")
-
+    
+    if ((patch_shape_x!=img_shape_x) or (patch_shape_y!=img_shape_y)) and sampling_method == "deterministic":
+        raise NotImplementedError("patchified generation not implemented yet for deterministic sampling")
```
Comment (mnabian) [inline examples/generative/corrdiff/generate.py:130]: Could you please make an issue for this?

File: examples/generative/corrdiff/shell_scripts/example_patch_train.sh [hunk 10]
```diff
@@ -0,0 +1 @@
+RUN_CMD="python3 -u train.py patch_shape_x=64 patch_shape_y=64 patch_num=10 batch=2560 batch_gpu=20 precond=resloss arch=ddpmpp-cwb outdir="./output_test""
```
Comment (mnabian) [inline examples/generative/corrdiff/shell_scripts/example_patch_train.sh:1]: Please note that config file names have changed and you need to pass `--config-name`. See the README here: https://github.com/NVIDIA/modulus/blob/main/examples/generative/corrdiff/README.md

File: examples/generative/corrdiff/train.py (hunk 7 above)
Comment (mnabian) [reply examples/generative/corrdiff/train.py:95]: yeah sounds good.

File: examples/generative/corrdiff/generate.py [hunk 11]
```diff
@@ -86,6 +85,10 @@ def main(cfg: DictConfig) -> None:
     out_channels = getattr(cfg, "out_channels", [0, 17, 18, 19])
     img_shape_x = getattr(cfg, "img_shape_x", 448)
     img_shape_y = getattr(cfg, "img_shape_y", 448)
+    patch_shape_x = getattr(cfg, "patch_shape_x", 448)
```
Comment (mnabian) [inline examples/generative/corrdiff/generate.py:88]: I wonder if we need to add a shape check for patches. Patch shapes cannot be any arbitrary number, right?

File: examples/generative/corrdiff/training/dataset.py [hunk 12]
```diff
@@ -496,7 +496,7 @@ def get_zarr_dataset(
         "v1": get_target_normalizations_v1,
         "v2": get_target_normalizations_v2,
     }[normalization]
-    logging.info("Normalization:", get_target_normalization)
+    logger.info(f"Normalization: {normalization}")
```
Comment (mnabian) [inline examples/generative/corrdiff/training/dataset.py:499]: Please merge main into your branch. This is already fixed.

File: examples/generative/corrdiff/training/training_loop.py [hunk 13]
```diff
@@ -154,6 +157,10 @@ def training_loop(
     )
 
     img_in_channels = len(in_channels)  # noise + low-res input
+
+    # interpolated global map added for patch-based resdiff
```
Comment (mnabian) [inline examples/generative/corrdiff/training/training_loop.py:161]: Change resdiff to corrdif

File: examples/generative/corrdiff/generate.py (hunk 11 above)
Comment (tge25) [reply examples/generative/corrdiff/generate.py:93]: A shape check for patches has been implemented, but the current check may not cover all possible scenarios comprehensively.

**Review by mnabian — APPROVED** [review]
